OutOfSchool/OutOfSchool.WebApi/Controllers/ChatWorkshopController.cs

```diff
@@ -56,30 +56,179 @@ public class ChatWorkshopController : ControllerBase
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public Task<IActionResult> GetProvidersRoomsAsync()
+            => this.GetUsersRoomsAsync(providerId => roomService.GetByProviderIdAsync(providerId));
+
+        /// <summary>
+        /// Set current date and time for all not read chat messages for parent in the specified chat room.
+        /// </summary>
+        /// <param name="id">ChatRoom's Id.</param>
+        /// <returns>Number of successfully updated items.</returns>
+        [HttpPatch("parent/chatrooms/{id}/read")]
+        [Authorize(Roles = "parent")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(int))]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public Task<IActionResult> SetReadDatetimeInMessagesForParentAsync([Range(1, long.MaxValue)] long id)
+            => this.SetReadDatetimeInMessagesForCurrentUserAsync(id, this.IsParentAChatRoomParticipantAsync);
+
+        /// <summary>
+        /// Set current date and time for all not read chat messages for provider in the specified chat room.
+        /// </summary>
+        /// <param name="id">ChatRoom's Id.</param>
+        /// <returns>Number of successfully updated items.</returns>
+        [HttpPatch("provider/chatrooms/{id}/read")]
+        [Authorize(Roles = "provider")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(int))]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public Task<IActionResult> SetReadDatetimeInMessagesForProviderAsync([Range(1, long.MaxValue)] long id)
+            => this.SetReadDatetimeInMessagesForCurrentUserAsync(id, this.IsProviderAChatRoomParticipantAsync);
+
+        private Task<bool> IsParentAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsParentOwnerAsync(userId, chatRoom.ParentId);
+        }
+
+        private Task<bool> IsProviderAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsWorkshopOwnerAsync(userId, chatRoom.WorkshopId);
+        }
+
+        private string GetUserId()
```

Comment:
`=> HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub)
                ?? throw new AuthenticationException($"Can not get user's claim {nameof(IdentityResourceClaimsTypes.Sub)} from HttpContext.");`
?
But we will throw that exception instead of return an HttpError, is it Ok ?

Reply:
In each controller method, we catch `AuthenticationException` and log it, and return BadRequest to the user with the neutral message "Can not get some user's claims. Please check your authentication or contact technical support."